lib/middleware-common/src/metering.rs

```diff
@@ -1,37 +1,118 @@
+/// Metering is a compiler middleware that calculates the cost of WebAssembly instructions at
+/// compile time and will count the cost of executed instructions at runtime. Within the Metering
+/// functionality, this instruction cost is called `points`.
 ///
-/// The Metering struct takes a `limit` parameter which is the maximum number of points which can be
-/// used by an instance during a function call. If this limit is exceeded, the function call will
-/// trap. Each instance has a `points_used` field which can be used to track points used during
-/// a function call and should be set back to zero after a function call.
+/// Each instance has an `exec_limit` which is the maximum number of points which can be used by
+/// the instance during a function call. If this limit is exceeded, the function call will trap.
+/// Each instance has a `points_used` field which can be used to track points used during a
+/// function call and should be set back to zero after a function call.
 ///
 /// Each compiler backend with Metering enabled should produce the same cost used at runtime for
 /// the same function calls so we can say that the metering is deterministic.
-///
 pub struct Metering {
-    limit: u64,
-    current_block: u64,
+    injections: Vec<Injection>,
+    current_block_injections: Vec<Injection>,
+    current_block_cost: u64,
 }
 
 impl Metering {
-    pub fn new(limit: u64) -> Metering {
+    pub fn new() -> Metering {
         Metering {
-            limit,
-            current_block: 0,
+            injections: Vec::new(),
+            current_block_injections: Vec::new(),
+            current_block_cost: 0,
+        }
+    }
+
+    fn set_costs<'a, 'b: 'a>(&mut self) {
+        for inj in &mut self.current_block_injections {
+            inj.check += self.current_block_cost;
+        }
+        // Set add of the last injection
+        if self.current_block_injections.len() > 0 {
+            let last_idx = self.current_block_injections.len() - 1;
+            self.current_block_injections[last_idx] = Injection {
+                add: self.current_block_cost,
+                check: 0,
+            };
+        }
+        self.current_block_cost = 0;
+    }
+
+    fn begin<'a, 'b: 'a>(&mut self) {
+        self.set_costs();
+        self.current_block_injections
+            .push(Injection { add: 0, check: 0 });
+    }
+    fn end<'a, 'b: 'a>(&mut self) {
+        self.set_costs();
+        self.injections.append(&mut self.current_block_injections);
+    }
+
+    fn inject_metering<'a, 'b: 'a>(&self, sink: &mut EventSink<'a, 'b>) {
+        let prev: Vec<Event> = sink.buffer.drain(..).collect();
+        let mut inj_idx: usize = 1;
+        for ev in prev {
+            match ev {
+                Event::Internal(FunctionBegin(_)) => {
+                    sink.push(ev);
+                    self.injections[0].inject(sink);
+                }
+                Event::Wasm(&ref op) | Event::WasmOwned(ref op) => match *op {
```

Comment:
⛳️ 
I believe `&ref op` is exactly the same as `op` I think, `ref` makes it a reference and `&` pattern matches the reference away

Reply:
I thought so too but I rant into issues when I used any different combination here. I think the compiler requires the `&ref op` on `Event::Wasm` to match `Event::WasmOwned(ref op)` but I can't precisely articulate why.

This exact match pattern was used in the original metering code.

In this particular branch I don't believe I will ever see any `Event::Wasm` as everything is `Wasm::Owned` once it is passed to the middleware in my current refactor of `codegen`, but I didn't want to assume that would always be the case in the middleware.

Reply:
Okay, that makes sense -- I find it easy to believe that when matching on multiple things to the same body that the types need to line up exactly and somehow that this makes it work 🤷‍♂ .  Thanks for looking into it!